test/jdk/tools/jlink/TaskHelperTest.java

```diff
@@ -0,0 +1,148 @@
+
+import jdk.tools.jlink.internal.PluginRepository;
+import jdk.tools.jlink.internal.TaskHelper;
+import jdk.tools.jlink.internal.TaskHelper.Option;
+import jdk.tools.jlink.internal.TaskHelper.OptionsHelper;
+import jdk.tools.jlink.plugin.Plugin;
+import jdk.tools.jlink.plugin.ResourcePool;
+import jdk.tools.jlink.plugin.ResourcePoolBuilder;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import jdk.tools.jlink.internal.TaskHelper.BadArgs;
+
+/*
+ * @test
+ * @summary Test TaskHelper option parsing
+ * @bug 8303884
+ * @modules jdk.jlink/jdk.tools.jlink.internal
+ *          jdk.jlink/jdk.tools.jlink.plugin
+ * @run junit TaskHelperTest
+ */
+public class TaskHelperTest {
+    private static TaskHelper taskHelper;
+    private static OptionsHelper<TaskHelperTest> optionsHelper;
+
+    private static final List<Option<TaskHelperTest>> OPTIONS = List.of(
+        new Option<>(true, (task, opt, arg) -> {
+            System.out.println(arg);
+            argValue = arg;
+        }, true, "--main-expecting"),
+        new Option<>(false, (task, opt, arg) -> {
+        }, true, "--main-no-arg")
+    );
+
+    private static String argValue;
+
+    public static class TestPluginWithRawOption implements Plugin {
+        @Override
+        public ResourcePool transform(ResourcePool in, ResourcePoolBuilder out) {
+            return out.build();
+        }
+
+        @Override
+        public boolean hasArguments() {
+            return true;
+        }
+
+        @Override
+        public boolean hasRawArgument() {
+            return true;
+        }
+
+        @Override
+        public String getName() {
+            return "raw-arg-plugin";
+        }
+
+        @Override
+        public void configure(Map<String, String> config) {
+            config.forEach((k, v) -> {
+                System.out.println(k + " -> " + v);
+            });
+            var v = config.get(getName());
+            if (v == null)
+                throw new AssertionError();
+            argValue = v;
+        }
+    }
+
+    @BeforeAll
+    public static void setup() {
+        taskHelper = new TaskHelper(TaskHelper.JLINK_BUNDLE);
+        optionsHelper = taskHelper.newOptionsHelper(TaskHelperTest.class, OPTIONS.toArray(Option[]::new));
+        PluginRepository.registerPlugin(new TestPluginWithRawOption());
+    }
+
+    @Test
+    public void testGnuStyleOptionAsArgValue() throws TaskHelper.BadArgs {
+        var validFormats = new String[][] {
+            { "--main-expecting=--main-no-arg", "--main-no-arg" },
+            { "--main-expecting", "--main-no-arg --list", "--main-no-arg"},
+            { "--main-expecting", " --main-no-arg", "--main-no-arg" },
+            { "--raw-arg-plugin=--main-no-arg", "--main-no-arg" },
+            { "--raw-arg-plugin", "--main-no-arg --list", "--main-no-arg"},
+            { "--raw-arg-plugin", " --main-no-arg", "--main-no-arg" },
+        };
```

Comment:
Line 126 validate the argument is processed.

Reply:
I expect it should check for the expected value rather than just `--main-no-arg` to catch bug like if the value contains multiple values but only returns one.